### DISCUSSION TITLE
Getting horrible results with thiebaud_xl_openpose

### DISCUSSION THREAD

**Discussion by dayunbao** [Mikubill/sd-webui-controlnet#2153] [Q&A]
I just upgraded my AWS EC2 instance type to a `g5.xlarge` so it can better handle SD XL.  The speed up I got was impressive.  Anyway, I did two generations to compare the quality of the images when using `thiebaud_xl_openpose` and when not using it.  

When not using it the results are beautiful:

![00046-1844297559](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/f1738fee-c0e3-4bf9-b557-51fc6092ea7c)
![00047-1844297560](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/aef0300b-b76b-49d7-98ef-12d763f08bc9)
![00048-1844297561](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/ef27b18b-025e-483f-ae7c-a04668902471)
![00049-1844297562](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/a45ba76a-d784-4d77-b609-7b72c881bdd2)

But when I use it, the results are horrible:
![00046-1844297559](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/c0299cf0-9cb3-4f83-8748-7c10864a62bb)
![00047-1844297560](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/58a38cb1-fa0d-4d5e-ba2a-e9ac1c76848f)
![00048-1844297561](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/1aa6ebd4-7ea8-4a5c-a661-252f55bb5aa1)
![00049-1844297562](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/2054ed39-9b0f-49cb-ba0a-e5eb9bc084f0)

Is anyone else getting similar results?  I make use of the OpenPose model extensively, but these results make it unusable.

Here are the prompts and settings I used.  

I started the server with the following command line arguments: `webui.sh --xformers --no-half-vae`

Without `thiebaud_xl_openpose`:
> A (humanoid: 1.5) (samurai:1.2) cat (standing:1.3) in a cyberpunk alleyway lit by neon signs
Steps: 50, Sampler: Euler a, CFG scale: 7, Seed: 1844297559, Size: 1024x1024, Model hash: 31e35c80fc, Model: sd_xl_base_1.0, VAE hash: 63aeecb90f, VAE: sdxl_vae.safetensors, Version: v1.6.0-1-g0562456f

With `thiebaud_xl_openpose`:
> A (humanoid: 1.5) (samurai:1.2) cat (swinging:1.3) a sword in a cyberpunk alleyway lit by neon signs
Steps: 50, Sampler: Euler a, CFG scale: 7, Seed: 1844297559, Size: 1024x1024, Model hash: 31e35c80fc, Model: sd_xl_base_1.0, VAE hash: 63aeecb90f, VAE: sdxl_vae.safetensors, ControlNet 0: "Module: dw_openpose_full, Model: thibaud_xl_openpose [c7b9cadd], Weight: 1, Resize Mode: Crop and Resize, Low Vram: True, Processor Res: 512, Guidance Start: 0, Guidance End: 1, Pixel Perfect: True, Control Mode: Balanced", Version: v1.6.0-1-g0562456f

**Comment by dayunbao** [comment]
I just realized that I forgot to include `--no-half-controlnet`.  So I just did that and tried again.  The results aren't much better:

![00066-3057226455](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/d1674e94-8f38-478f-8d7d-57f151eeeea1)
![00067-3057226456](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/975e0441-d817-49b2-8b7d-4de41547b9e8)
![00068-3057226457](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/94268b45-f0e8-463d-b5bb-70e3311f40c5)
![00069-3057226458](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/d0066753-93d1-4079-bdf3-3734145baa5c)

The prompt is slightly different, and I used a different sampler:
> A (humanoid: 1.5) (samurai:1.2) (black: 1.2) cat (standing:1.3) (holding a katana in one hand: 1.3) in a cyberpunk alleyway lit by neon signs
Negative prompt: Extra limbs, extra fingers, deformed hands, low quality,
Steps: 50, Sampler: UniPC, CFG scale: 7, Seed: 3057226458, Size: 1024x1024, Model hash: 31e35c80fc, Model: sd_xl_base_1.0, VAE hash: 63aeecb90f, VAE: sdxl_vae.safetensors, ControlNet 0: "Module: dw_openpose_full, Model: thibaud_xl_openpose [c7b9cadd], Weight: 1, Resize Mode: Crop and Resize, Low Vram: True, Processor Res: 512, Guidance Start: 0, Guidance End: 1, Pixel Perfect: True, Control Mode: Balanced", Version: v1.6.0-1-g0562456f

**Reply by dayunbao** [reply]
The results with other OpenPose models aren't much better, in fact, they don't seem to even be using the pose:

t2i-adapter_diffusers_xl_openpose:
![00070-663127863](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/85063656-0e5f-4624-91b4-7cf62da04c2f)

kohya_controllllite_xl_openpose_anime_v2:
![00080-3818839314](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/e233c7c8-b74a-4112-b4a9-9021426c9a4f)

I used the `dw_openpose_full` preprocessor with each.  Should I have used something different?

**Reply by huchenlei** [reply]
I think none of the current SDXL controlnet openpose model is trained with hand/face keypoints data. So when you use these models, you should probably either use `openpose` preprocessor that does not include hand/face or manually remove hand/face using editor.

**Reply by dayunbao** [reply]
OK, but there is still something wrong.  The only SD XL OpenPose model that consistently recognizes the OpenPose body keypoints is `thiebaud_xl_openpose`.  But that model destroys all the images.  It's unusable.  So I've been using `kohya_controllllite_xl_openpose_anime_v2`, but that only seems to recognize the OpenPose body keypoints about 35-40% of the time, assuming it's recognizing them at all, and it's not just luck that the poses of some of the characters match the keypoints.  

An example.  This is the input:
![woman_leaning_openpose](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/493c07cb-b556-4876-8bb6-199dca81dc1b)

These are the results:
![result_1](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/fd0de8ec-3169-44f1-b3b8-d1721eca39c1)
![result_2](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/c9514548-1970-4210-9b50-88aab65f1bc9)

The output of the logs during image generation:
```
Oct 05 06:40:41: 2023-10-05 06:40:41,443 - ControlNet - INFO - Loading model from cache: kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb]
Oct 05 06:40:41: 2023-10-05 06:40:41,447 - ControlNet - INFO - Loading preprocessor: none
Oct 05 06:40:41: 2023-10-05 06:40:41,447 - ControlNet - INFO - preprocessor resolution = 512
Oct 05 06:40:41: 2023-10-05 06:40:41,483 - ControlNet - INFO - ControlNet Hooked - Time = 0.05967235565185547
```

Here's the results using `thiebaud_xl_openpose`:
![thiebaud_result_1](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/5a2a06c4-e0ac-4f22-b0ee-b5c3cd53e4f6)
![thiebaud_result_2](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/4107546f-564f-4a40-ab9d-1979aa9fc58b)


and the API logs:
```
Oct 05 06:44:40: 2023-10-05 06:44:40,247 - ControlNet - INFO - Loading model: thibaud_xl_openpose [c7b9cadd]
Oct 05 06:44:40: 2023-10-05 06:44:40,336 - ControlNet - INFO - Loaded state_dict from [/home//stable-diffusion-webui/extensions/sd-webui-controlnet/models/thibaud_xl_openpose.safetensors]
Oct 05 06:44:40: 2023-10-05 06:44:40,348 - ControlNet - INFO - controlnet_sdxl_config
Oct 05 06:44:47: 2023-10-05 06:44:47,437 - ControlNet - INFO - ControlNet model thibaud_xl_openpose [c7b9cadd] loaded.
Oct 05 06:44:47: 2023-10-05 06:44:47,509 - ControlNet - INFO - Loading preprocessor: none
Oct 05 06:44:47: 2023-10-05 06:44:47,509 - ControlNet - INFO - preprocessor resolution = 512
Oct 05 06:44:47: 2023-10-05 06:44:47,541 - ControlNet - INFO - ControlNet Hooked - Time = 7.52367377281189
```

Do I need to change some settings, or is `kohya_controllllite_xl_openpose_anime_v2` just not working?

**Reply by dayunbao** [reply]
I made a fresh call after restarting the server to make sure everything was being loaded the same way.  Looks like it is:
```
Oct 05 06:57:46: 2023-10-05 06:57:46,315 - ControlNet - INFO - Loading model: kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb]
Oct 05 06:57:46: 2023-10-05 06:57:46,915 - ControlNet - INFO - Loaded state_dict from [/home/stable-diffusion-webui/extensions/sd-webui-controlnet/models/kohya_controllllite_xl_openpose_anime_v2.safetensors]
Oct 05 06:57:47: 2023-10-05 06:57:47,440 - ControlNet - INFO - ControlNet model kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb] loaded.
Oct 05 06:57:47: 2023-10-05 06:57:47,447 - ControlNet - INFO - Loading preprocessor: none
Oct 05 06:57:47: 2023-10-05 06:57:47,447 - ControlNet - INFO - preprocessor resolution = 512
Oct 05 06:57:47: 2023-10-05 06:57:47,491 - ControlNet - INFO - ControlNet Hooked - Time = 1.3990898132324219
```

**Comment by dayunbao** [comment]
After running a very similar generation in both the web UI and in my app via the API, I think there's a bug somewhere in the API. 

![web_ui_controlnet_setting](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/5f5fc15a-b79c-464f-b967-1a50e1019ebe)
![web_ui_results](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/ebfcdadb-ea05-4843-8ae2-e526f018054f)


Web UI settings:
```
A cute Asian woman standing in a kitchen
Steps: 50, 
Sampler: DPM++ 3M SDE Karras, 
CFG scale: 7, 
Seed: 3178681427, 
Size: 1024x1024, 
Model hash: 31e35c80fc, 
Model: sd_xl_base_1.0, 
VAE hash: 63aeecb90f, 
VAE: sdxl_vae.safetensors, 

ControlNet 0: 
    "Module: openpose, 
    Model: kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb], 
    Weight: 1, 
    Resize Mode: Crop and Resize, 
    Low Vram: False, 
    Processor Res: 512, 
    Guidance Start: 0, 
    Guidance End: 1, 
    Pixel Perfect: False, 
    Control Mode: Balanced"

Version: v1.6.0-1-g0562456f
```

Request body:
```
{
    "prompt": "A cute Asian woman standing in a kitchen",
    "negative_prompt": "",
    "cfg_scale": 7,
    "sampler_name": "DPM++ 3M SDE Karras",
    "batch_size": 8,
    "seed": 0,
    "steps": 50,
    "alwayson_scripts": {
        "controlnet": {
            "args": [
                {
                    "input_image": "a base64 string",
                    "enabled": true,
                    "lowvram": false,
                    "pixel_perfect": false,
                    "module": "openpose",
                    "model": "kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb]",
                    "processor_res": 512,
                    "threshold_a": 64,
                    "threshold_b": 64,
                    "weight": 1,
                    "guidance_start": 0,
                    "guidance_end": 1,
                    "control_mode": 0,
                    "resize_mode": 1
                }
            ]
        }
    }
}
```

![set_up](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/cd77125d-3d13-4f98-a82f-3c8a53374c7b)

Results:
![result_1](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/902f8b9c-00de-4106-835f-127c819045e7)
![result_2](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/81eb4ee7-de03-47e1-812e-5c756a292b14)

**Comment by dayunbao** [comment]
Another test, since I realized I wasn't passing the image width and height when calling txt2img:

Request body:
```
{
    "prompt": "A cute Asian woman standing in a kitchen",
    "negative_prompt": "",
    "cfg_scale": 7,
    "sampler_name": "DPM++ 3M SDE Karras",
    "batch_size": 8,
    "seed": 0,
    "height": 1024,
    "width": 1024,
    "steps": 50,
    "alwayson_scripts": {
        "controlnet": {
            "args": [
                {
                    "input_image": "",
                    "enabled": true,
                    "lowvram": false,
                    "pixel_perfect": false,
                    "module": "openpose",
                    "model": "kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb]",
                    "processor_res": 512,
                    "threshold_a": 64,
                    "threshold_b": 64,
                    "weight": 1,
                    "guidance_start": 0,
                    "guidance_end": 1,
                    "control_mode": 0,
                    "resize_mode": 1
                }
            ]
        }
    }
}
```

I am passing the `input_image`, by the way.  I just didn't want to paste a huge base64 encoded string.

I used the same OpenPose keypoints:
![woman_leaning_openpose](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/d6fa0c40-9442-4dc3-86e8-81e9cdb1f501)

But it still seems that `kohya_controllllite_xl_openpose_anime_v2` is ignoring the keypoints:
![result_again_1](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/f86187e0-5cc2-489e-bab3-98fe33e7d483)
![result_again_2](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/8de73ca3-0410-45ad-b9f7-603140b4fb95)

**Comment by wandrzej** [comment]
overall most if not all openpose controlnets for SDXL are bad. They either ignore the pose, or hit the quality, or both. 
afaik there was a half decent one from T2I, but still not even close to what one could get with SD1.5 models.

**Reply by dayunbao** [reply]
I'll try the T2I model some more, thanks.

My co-founders and I were already thinking about on training our own OpenPose model for ControlNet.  Now it looks like we'll definitely have to,

**Reply by dayunbao** [reply]
Is this just a problem when using Automatic1111's web UI?  Does it work any better in ComfyUI?

**Comment by dayunbao** [comment]
There is a new OpnePose XL model available from the maker of `thiebaud_xl_openpose`:

https://huggingface.co/thibaud/controlnet-openpose-sdxl-1.0/tree/main

Maybe this one would work better? @lllyasviel do you think we could get support added for this model?

**Reply by dayunbao** [reply]
@lllyasviel I dropped that model into the appropriate folder, and tried using it since the yaml files aren't used any more.  The API log said it was loaded, but it didn't seem to work.  The images didn't respect the OpenPose keypoints at all.  I don't know if there is any special configuration that needs to be done before that model can be used, though.

**Reply by dedkamaroz** [reply]
I'm just wondering if you've had any updates about this issue?

**Reply by dayunbao** [reply]
No, unfortunately.  Unless someone has released new ControlNet OpenPose models for SD XL, we're all borked.  The problem seems to lie with the poorly trained models, not ControlNet or this extension.  So I'll close this.  If my startup is able to get funding, I'm planning on setting aside money specifically to train ControlNet OpenPose models.

**Reply by sdbds** [reply]
https://huggingface.co/bdsqlsz/qinglong_controlnet-lllite/
you can test my openpose model.
